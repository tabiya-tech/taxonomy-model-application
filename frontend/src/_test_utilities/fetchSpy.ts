export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
