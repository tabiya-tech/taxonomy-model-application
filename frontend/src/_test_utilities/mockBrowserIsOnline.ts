import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
