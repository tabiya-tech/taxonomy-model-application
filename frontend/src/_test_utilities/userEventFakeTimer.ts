import userEvent from "@testing-library/user-event";
import { act } from "@testing-library/react";
import { UserEvent } from "@testing-library/user-event/setup/setup";

async function setup(callback: (user: UserEvent) => Promise<void>) {
  // We need to use fake times to ensure the debounce timer is progressed,
  // Since the debounce timer is called after the user stops typing, we need to set the fake timers
  // before the user begins typing, and for this reason we need to set up the userEvent to advance the timers using the jest timers,
  // so that the await user.type() can be resolved.
  // User enters a model name
  jest.useFakeTimers();
  const userEventFakeTimer = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
  await callback(userEventFakeTimer);
  act(() => jest.runOnlyPendingTimers());
  jest.useRealTimers();
}

export async function typeDebouncedInput(element: Element, text: string) {
  await setup(async (user) => {
    await user.type(element, text);
  });
}

export async function clickDebouncedButton(element: Element) {
  await setup(async (user) => {
    await user.click(element);
  });
}
