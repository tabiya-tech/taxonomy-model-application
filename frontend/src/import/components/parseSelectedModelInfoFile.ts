import Papa from "papaparse";

interface CsvRow {
  UUIDHISTORY?: string;
  DESCRIPTION?: string;
}

type ModelInfoDetails = {
  UUIDHistory: string[];
  description: string;
};

const parseSelectedModelInfoFile = (file: File): Promise<ModelInfoDetails> => {
  return new Promise((resolve, reject) => {
    // Check if the file is a CSV
    if (file.type !== "text/csv" && file.type !== "application/vnd.ms-excel") {
      reject(new Error("File is not a CSV"));
      return;
    }

    Papa.parse<CsvRow>(file, {
      complete: (result) => {
        const records = result.data;

        // Check if the UUIDHistory column is found and not empty
        if (records.length === 0 || !records[0].UUIDHISTORY) {
          reject(new Error("UUIDHistory column not found or empty in the first row"));
          return;
        }

        // Split the UUIDHistory string into an array of UUIDs
        const uuidHistory: string[] = records[0].UUIDHISTORY.split("\n");
        const description = records[0].DESCRIPTION || "";

        resolve({
          UUIDHistory: uuidHistory,
          description: description,
        });
      },
      error: (error) => reject(new Error("Error parsing the file: " + error.message)),
      header: true,
      skipEmptyLines: "greedy", // Skips empty lines
      // preview: 1, // Only parse the first row of actual data, ignoring the header
    });
  });
};

export default parseSelectedModelInfoFile;
