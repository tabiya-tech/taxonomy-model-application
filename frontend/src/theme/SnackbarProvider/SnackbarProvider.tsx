import {
  SnackbarProvider as OriginalSnackbarProvider,
  SnackbarProviderProps as OriginalSnackbarProviderProps,
  useSnackbar,
  SnackbarKey,
  MaterialDesignContent,
} from "notistack";
import { IconButton, styled } from "@mui/material";
import CloseIcon from "@mui/icons-material/Close";
import * as React from "react";

const uniqueId = "bade4e92-baed-40f8-a098-0e745563e786";
export const DATA_TEST_ID = {
  SNACKBAR_CLOSE_BUTTON: `snackbar-close-button-${uniqueId}`,
};

const SnackbarCloseButton = (key: SnackbarKey) => {
  const { closeSnackbar } = useSnackbar();
  // Currently due to a bug in the notistack library, we cannot ref to the correct ARIA ids (see https://github.com/iamhosseindhv/notistack/issues/579)
  // Suing  the title attribute for accessibility, is not ideal. When multiple snackbars are open,
  // the title will be the same for all of them violating the WAG2A rule of unique id. See https://tabiya-tech.atlassian.net/browse/PLAT-129
  return (
    <IconButton
      title={"Close notification"}
      data-testid={DATA_TEST_ID.SNACKBAR_CLOSE_BUTTON}
      onClick={() => closeSnackbar(key)}
      sx={{
        margin: "0",
        padding: "0",
      }}
    >
      <CloseIcon
        sx={{
          margin: 0,
          padding: 0,
        }}
      />
    </IconButton>
  );
};

const StyledMaterialDesignContent = styled(MaterialDesignContent)(({ theme }) => ({
  "&.notistack-MuiContent": {
    // adjust the snackbar content to place the close button on the top right,
    // so that the button is not hidden in case multiple snackbars are open and overlap,
    // especially when the text is long
    display: "flex",
    flexDirection: "row",
    flexWrap: "nowrap",
    alignItems: "start",
    // ---
  },
  "&.notistack-MuiContent-success": {
    color: theme.palette.success.dark,
    backgroundColor: theme.palette.success.light,
    fontFamily: theme.typography.body1.fontFamily,
    fontSize: theme.typography.body1.fontSize,
    "& .MuiSvgIcon-root": {
      color: theme.palette.success.dark,
    },
  },
  "& #notistack-snackbar svg": {
    fontSize: theme.typography.h4.fontSize + "!important",
  },
  "&.notistack-MuiContent-error": {
    color: theme.palette.error.dark,
    backgroundColor: theme.palette.error.light,
    fontFamily: theme.typography.body1.fontFamily,
    fontSize: theme.typography.body1.fontSize,
    "& .MuiSvgIcon-root": {
      color: theme.palette.error.dark,
    },
  },
  "&.notistack-MuiContent-warning": {
    color: theme.palette.warning.dark,
    backgroundColor: theme.palette.warning.light,
    fontFamily: theme.typography.body1.fontFamily,
    fontSize: theme.typography.body1.fontSize,
    "& .MuiSvgIcon-root": {
      color: theme.palette.warning.dark,
    },
  },
  "&.notistack-MuiContent-info": {
    color: theme.palette.info.dark,
    backgroundColor: theme.palette.info.light,
    fontFamily: theme.typography.body1.fontFamily,
    fontSize: theme.typography.body1.fontSize,
    "& .MuiSvgIcon-root": {
      color: theme.palette.info.dark,
    },
  },
}));

const SnackbarProvider: React.FC<OriginalSnackbarProviderProps> = ({
  children,
  ...props
}: Readonly<OriginalSnackbarProviderProps>) => {
  const notistackOptions: OriginalSnackbarProviderProps = {
    maxSnack: 10,
    autoHideDuration: null,
    transitionDuration: { enter: 500, exit: 500 },
    anchorOrigin: {
      vertical: "top",
      horizontal: "right",
    },
    Components: {
      success: StyledMaterialDesignContent,
      error: StyledMaterialDesignContent,
      warning: StyledMaterialDesignContent,
      info: StyledMaterialDesignContent,
    },
    dense: false,
    style: {
      // limit the width of the snackbar so that long text is wrapped in multiple lines
      maxWidth: "600px",
      whiteSpace: "normal",
      // ---
    },
    ...props,
  };

  return (
    <OriginalSnackbarProvider {...notistackOptions} action={SnackbarCloseButton}>
      {children}
    </OriginalSnackbarProvider>
  );
};

export default SnackbarProvider;
export { useSnackbar };
export type { VariantType } from "notistack";
