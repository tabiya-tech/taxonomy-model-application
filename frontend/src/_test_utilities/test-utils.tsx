// Based on https://testing-library.com/docs/react-testing-library/setup/

import React, { ReactElement } from "react";
import {render, renderHook, RenderOptions} from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/providers";
import {AuthProvider} from "../app/providers/AuthProvider";

type CustomRenderOptions = {
  useAuthProvider?: boolean
}

const AllTheProviders = ({ children, useAuthProvider = false }: { children: React.ReactNode } & CustomRenderOptions) => {
  const AuthProviderComponent = useAuthProvider ? AuthProvider : React.Fragment;

  return (
      <IsOnlineProvider>
        <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
            <AuthProviderComponent>
              <SnackbarProvider>{children}</SnackbarProvider>
            </AuthProviderComponent>
          </ThemeProvider>
      </IsOnlineProvider>
  );
};

type TOptions = Omit<RenderOptions, "wrapper"> & CustomRenderOptions
const customRender = (ui: ReactElement, options?: TOptions) =>
  render(ui, { wrapper: ({ children }) => <AllTheProviders useAuthProvider={options?.useAuthProvider}>{children}</AllTheProviders>, ...options });

const customRenderHook = (hook: () => any, options?: TOptions) =>
  renderHook(hook, { wrapper: ({ children }) => <AllTheProviders useAuthProvider={options?.useAuthProvider}>{children}</AllTheProviders>, ...options });

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
