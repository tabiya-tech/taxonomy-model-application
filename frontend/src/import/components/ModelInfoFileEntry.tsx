import * as React from "react";
import { ChangeEvent, useState } from "react";
import { Chip } from "@mui/material";
import debounce from "lodash.debounce";
import { DEBOUNCE_INTERVAL } from "./debouncing";
import { AddCircleOutlined, RemoveCircleOutlined } from "@mui/icons-material";
import parseSelectedModelInfoFile from "./parseSelectedModelInfoFile";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";

export interface ModelInfoFileEntryProps {
  notifyUUIDHistoryChange?: (newUUIDHistory: string[]) => void;
  notifyOnDescriptionChange?: (description: string) => void;
}

const uniqueId = "16c54d56-b091-48ce-826a-c721b0c3643d";

export const DATA_TEST_ID = {
  FILE_ENTRY: `modelInfo-file-entry-${uniqueId}`,
  FILE_INPUT: `modelInfo-file-input-${uniqueId}`,
  SELECT_FILE_BUTTON: `modelInfo-select-file-button-${uniqueId}`,
  REMOVE_SELECTED_FILE_BUTTON: `modelInfo-remove-selected-file-button-${uniqueId}`,
};
export const modelInfoFileType = "MODEL_INFO";
export const modelInfoFileTypeName = "Model info";
/**
 * Represent a file entry of the model info type and a selected file
 * The data-filetype attribute is used to identify the file type
 * @constructor
 */

export const ModelInfoFileEntry = (props: Readonly<ModelInfoFileEntryProps>) => {
  const [selectedFile, setSelectedFile] = useState<File | null>(null);
  const { enqueueSnackbar } = useSnackbar();

  const fileChangedHandler = async (e: ChangeEvent<HTMLInputElement>) => {
    const newFile: File = e.target?.files![0];
    if (newFile) {
      await updateSelectedFile(newFile);
    }
  };

  const fileRemovedHandler = async () => {
    await updateSelectedFile(null);
  };

  const notifyOnModelInfoChanges = (details: Awaited<ReturnType<typeof parseSelectedModelInfoFile>>) => {
    if (props.notifyUUIDHistoryChange) {
      props.notifyUUIDHistoryChange(details.UUIDHistory);
    }

    if (props.notifyOnDescriptionChange) {
      props.notifyOnDescriptionChange(details.description);
    }
  };

  const updateSelectedFile = async (file: File | null) => {
    setSelectedFile(file); // update internal state
    // notify the parent component if it has provided handler
    if (file) {
      try {
        const modelInfoDetails = await parseSelectedModelInfoFile(file);
        notifyOnModelInfoChanges(modelInfoDetails);
      } catch (e) {
        notifyOnModelInfoChanges({
          UUIDHistory: [],
          description: "",
        });

        console.error(e);
        enqueueSnackbar(`Error parsing file: ${file.name}. Please review the file and try again.`, {
          variant: "error",
        });
        setSelectedFile(null);
      }
    } else {
      notifyOnModelInfoChanges({
        UUIDHistory: [],
        description: "",
      });
    }
  };

  const debounceFileRemoveHandler = debounce(fileRemovedHandler, DEBOUNCE_INTERVAL);

  return (
    <div data-filetype={modelInfoFileType} data-testid={DATA_TEST_ID.FILE_ENTRY}>
      {selectedFile ? (
        <Chip
          color="secondary"
          aria-label={`Remove modelInfo csv file`}
          data-testid={DATA_TEST_ID.REMOVE_SELECTED_FILE_BUTTON}
          onClick={debounceFileRemoveHandler}
          icon={<RemoveCircleOutlined />}
          label={`${modelInfoFileTypeName}: ${selectedFile.name}`}
        />
      ) : (
        <div>
          <input
            id={`${uniqueId}-model-info`}
            type="file"
            style={{ display: "none" }}
            accept=".csv"
            data-testid={DATA_TEST_ID.FILE_INPUT}
            onChange={fileChangedHandler}
            data-filetype={modelInfoFileType}
          />
          <Chip
            color="primary"
            aria-label={`Add ${modelInfoFileType} csv file`}
            data-testid={DATA_TEST_ID.SELECT_FILE_BUTTON}
            onClick={() => document.getElementById(`${uniqueId}-model-info`)!.click()}
            icon={<AddCircleOutlined />}
            label={modelInfoFileTypeName}
          />
        </div>
      )}
    </div>
  );
};
export default ModelInfoFileEntry;
