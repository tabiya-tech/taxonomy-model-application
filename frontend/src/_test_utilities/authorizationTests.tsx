import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
