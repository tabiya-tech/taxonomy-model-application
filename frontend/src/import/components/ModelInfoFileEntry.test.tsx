// mute the console
import "src/_test_utilities/consoleMock";

import { fireEvent, render, screen, waitFor } from "src/_test_utilities/test-utils";
import { clickDebouncedButton } from "src/_test_utilities/userEventFakeTimer";
import ModelInfoFileEntry, { DATA_TEST_ID, modelInfoFileType, modelInfoFileTypeName } from "./ModelInfoFileEntry";
import parseSelectedModelInfoFile from "./parseSelectedModelInfoFile";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";

// mock the snackbar
jest.mock("src/theme/SnackbarProvider/SnackbarProvider", () => {
  const actual = jest.requireActual("src/theme/SnackbarProvider/SnackbarProvider");
  return {
    ...actual,
    __esModule: true,
    useSnackbar: jest.fn().mockReturnValue({
      enqueueSnackbar: jest.fn(),
      closeSnackbar: jest.fn(),
    }),
  };
});

describe("ModelInfoFileEntry render tests", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("Should render the modelInfoFileEntry component", () => {
    // When ModelInfoFileEntry is rendered
    render(<ModelInfoFileEntry />);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the component to be rendered
    const fileEntry = screen.getByTestId(DATA_TEST_ID.FILE_ENTRY);
    expect(fileEntry).toBeInTheDocument();
    expect(fileEntry).toMatchSnapshot();
    // AND to have the correct file type
    expect(fileEntry).toHaveAttribute("data-filetype");

    // AND expect file input's to have no files
    const fileInput: HTMLInputElement = screen.getByTestId(DATA_TEST_ID.FILE_INPUT);
    expect(fileInput.files).toHaveLength(0);
    // AND expect file input to have the file type
    expect(fileInput).toHaveAttribute("data-filetype", modelInfoFileType);

    // AND expect file trigger fab to be in the document
    const selectFileButton = screen.getByTestId(DATA_TEST_ID.SELECT_FILE_BUTTON);
    expect(selectFileButton).toBeInTheDocument();

    // AND expect file remover fab to not be in the document
    const fileRemoverButton = screen.queryByTestId(DATA_TEST_ID.REMOVE_SELECTED_FILE_BUTTON);
    expect(fileRemoverButton).not.toBeInTheDocument();

    // AND expect the SELECT_FILE_BUTTON to have the correct file type name
    expect(selectFileButton.textContent).toBe(modelInfoFileTypeName);
  });
});

describe("ModelInfoFileEntry action tests", () => {
  it("should selected file", async () => {
    // GIVEN some file
    const givenFile = new File([], "foo.csv", { type: "text/csv" });

    // When ModelInfoFileEntry is rendered
    render(<ModelInfoFileEntry />);
    // AND fileInput value is changed
    const fileInput: HTMLInputElement = screen.getByTestId(DATA_TEST_ID.FILE_INPUT);
    fireEvent.change(fileInput, { target: { files: [givenFile] } });

    // THEN expect file input to have 1 file
    expect(fileInput.files).toHaveLength(1);

    // AND expect file trigger fab to not be in the document
    const selectFileButton = screen.queryByTestId(DATA_TEST_ID.SELECT_FILE_BUTTON);
    expect(selectFileButton).not.toBeInTheDocument();

    // AND expect file remover fab to be in the document
    const fileRemoverFab = screen.queryByTestId(DATA_TEST_ID.REMOVE_SELECTED_FILE_BUTTON);
    expect(fileRemoverFab).toBeInTheDocument();
  });

  it("should remove selected file", async () => {
    // GIVEN some file
    const givenFile = new File([], "foo.csv", { type: "text/csv" });
    // When ModelInfoFileEntry is rendered
    render(<ModelInfoFileEntry />);
    // AND fileInput value has changed
    let fileInput: HTMLInputElement = screen.getByTestId(DATA_TEST_ID.FILE_INPUT);
    fireEvent.change(fileInput, { target: { files: [givenFile] } });
    expect(fileInput.files).toHaveLength(1);

    // WHEN fileRemoverFab is clicked
    const fileRemoverFab = screen.getByTestId(DATA_TEST_ID.REMOVE_SELECTED_FILE_BUTTON);
    await clickDebouncedButton(fileRemoverFab);

    // THEN expect file input to have an empty files
    // wait for the component's state to update
    await waitFor(() => {
      fileInput = screen.getByTestId(DATA_TEST_ID.FILE_INPUT);
    }, {});
    expect(fileInput.files).toHaveLength(0);
    // AND expect file trigger fab to be in the document
    const selectFileButton = screen.getByTestId(DATA_TEST_ID.SELECT_FILE_BUTTON);
    expect(selectFileButton).toBeInTheDocument();
  });

  it("should correctly notify the notifyUUIDHistoryChange handler when file is selected", async () => {
    // GIVEN some file
    const givenFile = new File([], "foo.csv", { type: "text/csv" });
    // AND a notification handler
    const givenMockNotification = jest.fn();

    // When ModelInfoFileEntry is rendered
    const givenParsedUUIDHistory = ["uuid1", "uuid2"];
    jest
      .spyOn(require("./parseSelectedModelInfoFile"), "default")
      .mockImplementationOnce(() => Promise.resolve(givenParsedUUIDHistory));
    render(<ModelInfoFileEntry notifyUUIDHistoryChange={givenMockNotification} />);
    // AND a file is chosen
    const fileInput = screen.getByTestId(DATA_TEST_ID.FILE_INPUT);
    fireEvent.change(fileInput, { target: { files: [givenFile] } });
    // THEN expect the parseSelectedModelInfoFile to have been called with the given file
    await waitFor(() => {
      expect(parseSelectedModelInfoFile).toHaveBeenCalledWith(givenFile);
    });
    // THEN expect notification to have been called with the given filetype and file
    expect(givenMockNotification).toHaveBeenCalledWith(givenParsedUUIDHistory);
  });

  it("should correctly notify the notifyUUIDHistoryChange handler when file is removed", async () => {
    // GIVEN some file
    const givenFile = new File([], "foo.csv", { type: "text/csv" });
    // AND a notification handler
    const givenMockNotification = jest.fn();

    // When ModelInfoFileEntry is rendered
    jest.spyOn(require("./parseSelectedModelInfoFile"), "default").mockImplementationOnce(() => Promise.resolve([]));
    render(<ModelInfoFileEntry notifyUUIDHistoryChange={givenMockNotification} />);
    // AND the given file has been selected
    let fileInput: HTMLInputElement = screen.getByTestId(DATA_TEST_ID.FILE_INPUT);
    fireEvent.change(fileInput, { target: { files: [givenFile] } });
    expect(fileInput.files).toHaveLength(1);

    // WHEN the remove selected file button is clicked
    const fileRemoverFab = screen.getByTestId(DATA_TEST_ID.REMOVE_SELECTED_FILE_BUTTON);
    // wait for the component's state to update
    await clickDebouncedButton(fileRemoverFab);
    // THEN expect notificationHandler to have been called with givenFileType and null,
    expect(givenMockNotification).toHaveBeenLastCalledWith([]);
  });

  it("should handle file changes even if notifyUUIDHistoryChange handler is not set", async () => {
    // GIVEN some file
    const givenAFile = new File([], "foo.csv", { type: "text/csv" });

    // When ModelInfoFileEntry is rendered without a notification handler
    render(<ModelInfoFileEntry />);
    // AND fileInput value has changed
    let fileInput: HTMLInputElement = screen.getByTestId(DATA_TEST_ID.FILE_INPUT);
    fireEvent.change(fileInput, { target: { files: [givenAFile] } });

    // THEN expect file input to have 1 files
    expect(fileInput.files).toHaveLength(1);

    // AND expect the fileRemoverButton to be in the document
    const fileRemoverButton = screen.getByTestId(DATA_TEST_ID.REMOVE_SELECTED_FILE_BUTTON);
    expect(fileRemoverButton).toBeInTheDocument();
  });

  it("should show a snackbar when the file is not parsed correctly", async () => {
    // GIVEN some file
    const givenFile = new File([], "foo.csv", { type: "text/csv" });
    // AND a notification handler
    const givenMockNotification = jest.fn();
    // AND a parseSelectedModelInfoFile that throws an error
    const givenError = new Error("parse error");
    jest
      .spyOn(require("./parseSelectedModelInfoFile"), "default")
      .mockImplementationOnce(() => Promise.reject(givenError));

    // When ModelInfoFileEntry is rendered
    render(<ModelInfoFileEntry notifyUUIDHistoryChange={givenMockNotification} />);
    // AND a file is chosen
    const fileInput = screen.getByTestId(DATA_TEST_ID.FILE_INPUT);
    fireEvent.change(fileInput, { target: { files: [givenFile] } });
    // THEN expect the parseSelectedModelInfoFile to have been called with the given file
    expect(parseSelectedModelInfoFile).toHaveBeenCalledWith(givenFile);
    // THEN expect the notification handler not to have been called
    expect(givenMockNotification).not.toHaveBeenCalled();
    // AND expect the error to have been logged to the console
    await waitFor(() => {
      expect(console.error).toHaveBeenCalledWith(givenError);
    });
    // THEN expect a snackbar to have been shown
    const { enqueueSnackbar } = useSnackbar();
    expect(enqueueSnackbar).toHaveBeenCalledWith(
      `Error parsing file: ${givenFile.name}. Please review the file and try again.`,
      { variant: "error" }
    );
  });
});
