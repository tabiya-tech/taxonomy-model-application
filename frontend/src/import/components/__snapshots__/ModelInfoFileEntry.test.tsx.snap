// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelInfoFileEntry render tests Should render the modelInfoFileEntry component 1`] = `
<div
  data-filetype="MODEL_INFO"
  data-testid="modelInfo-file-entry-16c54d56-b091-48ce-826a-c721b0c3643d"
>
  <div>
    <input
      accept=".csv"
      data-filetype="MODEL_INFO"
      data-testid="modelInfo-file-input-16c54d56-b091-48ce-826a-c721b0c3643d"
      id="16c54d56-b091-48ce-826a-c721b0c3643d-model-info"
      style="display: none;"
      type="file"
    />
    <div
      aria-label="Add MODEL_INFO csv file"
      class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-clickable MuiChip-clickableColorPrimary MuiChip-filledPrimary css-d8rrhm-MuiButtonBase-root-MuiChip-root"
      data-testid="modelInfo-select-file-button-16c54d56-b091-48ce-826a-c721b0c3643d"
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconMedium MuiChip-iconColorPrimary css-etbowg-MuiSvgIcon-root"
        data-testid="AddCircleOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
        />
      </svg>
      <span
        class="MuiChip-label MuiChip-labelMedium css-1dybbl5-MuiChip-label"
      >
        Model info
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
  </div>
</div>
`;
